DOC: Transfer the wiki FAQ page to Markdown documentation

jhlegarreta opened the pull request:
Transfer the wiki FAQ page to Markdown documentation: https://itk.org/Wiki/ITK/FAQ

Add the corresponding entry in the `Learn` section of the documentation index.

Take advantage of the transition to:
- Update URLs, including changing `http` to `https`, fixing broken URLs,
  etc.
- Link ITK and VTK classes to their Doxygen documentation pages.
- Highlight parts (e.g. class names) using Markdown syntax.
- Refactor the data format section to: include mesh and point set, and
  transform file format sections, mention other file formats potentially
  supported by remote modules. Add links to the file format reference
  pages, and the Doxygen documentation of the ITK IO classes involved.
- Make the BibTeX list style consistent, remove broken URLs pointing to
  old versions of the SW Guide or all pointing to the latest PDF
  available online, etc.
- Make the BibTeX list style consistent, remove broken URLs pointing to
  old versions of the SW Guide or all pointing to the latest PDF
  available online, etc.
- Remove outdated paragraph about libraries to build visualization
  applications.
- Change mentions to `ImageViewer` to `itkwidgets`.
- Prefer pointing to the Python `imageio` package over `ImageMagick`.
- Mention `itk-wasm dicom` to read a volume from a DICOM series.
- Reword the language binding section: remove Java and mention that
  further bindings are available through SimpleITK and itk-wasm.
- Remove the `What does \`Error opening hint file NOTFOUND\` mean?`
  section.
- Remove the language binding/wrapping section, as it is outdated: it
  refers to the current SWIG wrapping before it was integrated into the
  ITK repository.
- Remove mentions to section number and pages in the ITK SW Guide, as
  they have changed and may change in the future.
- Remove the emacs coding style section.
- Rewrite the Vim coding style section, as ITK's coding style is now
  enforced and/or checked automatically, and as ITK switched from
  `txx` to `hxx` suffices some time ago.
- Remove the empty `What is the `\`ITK_TEMPLATE_VISIBILITY_DEFAULT\`
  CMake option?` section.
- Remove the outdated `Platform-specific questions`section.
- Fix typos and grammar, and reword passages for better understanding.
- Change the title to `Frequently asked questions`.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,612 @@
+JPEG is that you can view them with almost any image viewer. It is
+possible to read/write 3D datasets as collections of 2D slices each one
+in an independent BMP, PNG or JPEG file by using the
+`ImageSerierReader` and [`itk::ImageSeriesWriter`](https://itk.org/Doxygen/html/classitk_1_1ImageSeriesWriter.html).
```
`ImageSerierReader` does not exist as a class.

thewtex replied:
I think it should be `ImageSeriesReader`.

jhlegarreta replied:
🤦. Done in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+is then use the VTKImageToImage adaptor in order to convert the
+[`vtkImageData`](https://vtk.org/doc/nightly/html/classvtkImageData.html) into an [`itk::Image`](https://itk.org/Doxygen/html/classitk_1_1Image.html).
+
+Supported VTK file formats can be found in https://docs.vtk.org/en/latest/supported_data_formats.html.
```
Updated VTK file formats to point to:
https://docs.vtk.org/en/latest/supported_data_formats.html
instead of
https://itk.org/Wiki/VTK_FAQ#What_image_file_formats_can_VTK_read_and_write.3F
and
https://itk.org/Wiki/VTK_FAQ#What_3D_file_formats_can_VTK_import_and_export.3F

jhlegarreta commented on Documentation/docs/learn/faq.md:
````diff
@@ -0,0 +1,611 @@
+VTK_WRAP_HINTS:FILEPATH=/home/johndoe/src/VTK/Wrapping/hints
+```
+
+### How do I use ITK with WrapITK?
````
Not sure what `WrapITK` is.

thewtex replied:
It is the current SWIG wrapping before it was integrated into the ITK repository. I think we can remove all these wrapping sections since they are out of date.

jhlegarreta replied:
Removed in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+in case of desperation.
+
+1. Use the [`itk::ImageSeriesReader`](https://itk.org/Doxygen/html/classitk_1_1ImageSeriesReader.html) in combination with the
+   `DicomSeriesFileNames`. For a full example on how to do this,
```
`DicomSeriesFileNames` does not exist as a class name

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+### How do I configure Emacs for ITK coding style?
+
+You can add the following codes into your `.emacs` file:
```
Not sure about this section. I do not use emacs.

thewtex replied:
I think it is likely no longer correct since our clang-format style update.

jhlegarreta replied:
Removed in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
````diff
@@ -0,0 +1,611 @@
+au BufNewFile,BufRead *.txx set filetype=cpp
+```
+
+### What is the `ITK_TEMPLATE_VISIBILITY_DEFAULT` CMake option?
````
Section is empty: is it relevant?

thewtex replied:
It can be removed.

jhlegarreta replied:
Removed in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+More information on [ITK File Formats](https://itk.org/Wiki/ITK_File_Formats) is
+available, including data types supported.
+
+A very nice tool for viewing all of these images is [ITKApps/ImageViewer](https://github.com/InsightSoftwareConsortium/ITKApps/tree/master/ImageViewer).
```
All mentions to the `ImageViewer` should be maybe changed to the ITK/VTK WebJS viewer?

jhlegarreta replied:
Done in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+High Bit       = 15
+
+   |<------------------------- pixel ------------------------->|
+    ______________ ______________ ______________ ______________
```
Changed span of `pixel`  dashes in here to cover the whole length. Not sure if the high bit was meant to be included.

jhlegarreta commented on Documentation/docs/learn/faq.md:
````diff
@@ -0,0 +1,611 @@
+```
+//Where can the hints file be found
+VTK_WRAP_HINTS:FILEPATH=/home/johndoe/src/VTK/Wrapping/hints
+```
````
Not sure if this section is still relevant.

thewtex replied:
It should be removed.

jhlegarreta replied:
Removed in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+It suffices to tell Vim to use such configuration file.
+
+You can also use CMake indentation and syntax file which are available
+on www.vim.org. This was contributed by Andy Cedilnik while at Kitware.
```
Not sure if this is still relevant.

thewtex replied:
This paragraph can be removed.

jhlegarreta replied:
Removed.

jhlegarreta commented on Documentation/docs/learn/faq.md:
````diff
@@ -0,0 +1,611 @@
+```
+syntax on
+au BufNewFile,BufRead *.txx set filetype=cpp
+```
````
Not sure if this is still relevant.

thewtex replied:
It can be removed -- we changed .txx to .hxx.

jhlegarreta replied:
Removed in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+   - SimpleITK is great, but it does not support all filters.
+   - A list of available filters can be found
+     [here](http://www.itk.org/SimpleITKDoxygen/html/Filter_Coverage.html)
+     (valid Sept 2012).
```
The link does not exist. This should be revisited.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+   - An example can be found [here](https://itk.org/Wiki/ITK/Using_ITK_from_.NET).
+
+NOTE: `WrapITK` does not currently support C# language bindings, but could
+be made to do so.
```
Not sure what this is.

jhlegarreta replied:
Removed the paragraph.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+WrapITK allows you to use ITK from other languages such as Python, Java, or
+Ruby). See the page about [ITK wrappings](https://itk.org/Wiki/ITK/Wrapping).
+
+### How do I use ITK from C# .NET?
```
This whole section needs revisiting.

jhlegarreta replied:
The section was removed.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+Currently ITK supports the following language bindings:
+
+- Python
+- Java
```
Should be revisited: not sure if we still support Java or whether this goes through SimpleITK.

thewtex replied:
We have Python language bindings, but Java-SWIG is not maintained.

We can point people to itk-wasm and SimpleITK.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+### How do I use ITK with WrapITK?
+
+WrapITK allows you to use ITK from other languages such as Python, Java, or
+Ruby). See the page about [ITK wrappings](https://itk.org/Wiki/ITK/Wrapping).
```
Needs to be re-written based on SimpleITK probably? I am not all that familiar with all the possibilities.

jhlegarreta replied:
Removed the whole section in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+   (with the motto: "MIND is not DICOM"). This application loads
+   DICOM files over the network and export them in MetaImage format.
+   This application is open source and it is available at:
+   http://www.jeffro.net/mind/ and http://caddlab.rad.unc.edu/software/MIND/.
```
Links are broken. Is this relevant? @aylward ?

jhlegarreta commented:
Sorry for being active only now; could not devote time to it during the day.

A few general notes:
- Used bare Markdown. Not sure if we are using another flavor.
- Did not add anchors to the titles.
- Could not find the 3rd edition (2013) ITK SWG book ISBN, so the note is empty.
- Changed `InsightApplications` to `ITKApps`: both were mentioned. Not sure if they are the same.
- Kept pointing examples to ITK SWG examples, rather than ITK Sphinx.
- IMO, the BibTeX entries of previous versions should go into a historical section, and we should always point to the latest edition.
  And all BibTeX files should be at a single place (there are some at the ISC website, some in ITK, there is a How to cite in the ITK README, etc. IIRC) to avoid inconsistencies.
- Links to other wiki pages that need to be transitioned:
https://itk.org/Wiki/ITK_File_Formats
https://itk.org/Wiki/FDA_Guidelines_for_Software_Development
https://itk.org/Wiki/ITK/Wrapping
https://itk.org/Wiki/ITK/Using_ITK_from_.NET
https://itk.org/Wiki/ITK/Testing
https://itk.org/Wiki/ITK_Hounsfield_Units
https://itk.org/Wiki/ITK/DICOM
https://itk.org/Wiki/ITK/Third_Party_Applications
- Broken links that need to be fixed for which I did not find a replacement:
http://www.itk.org/SimpleITKDoxygen/html/Filter_Coverage.html
http://www.itk.org/CourseWare/Training/GettingStarted-II.pdf: closest is https://itk.org/Wiki/ITK/Course_Ware
http://www.itk.org/CourseWare/Training/RegistrationMethodsOverview.pdf

- There are different types of wiki page deprecations, e.g.
https://itk.org/Wiki/ITKContribute
https://itk.org/Wiki/ITK/Examples/Iterators/ImageRegionIterator (content removed)

which one should be used?

- We would need to pass a Markdown linter to all these files
- Did not change line length after automatic conversion (72 vs 79).

A few in-line notes.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+The following are alternative options for reading DICOM series. The
+first one in the list is the recommended option. The others are provided
+in case of desperation.
+
```
I think option 0 here should be @itk-wasm dicom, which has the broadest support and is the easiest to use. [python docs](https://itk-wasm-dicom-python-docs.on.fleek.co/), [js docs](https://itk-wasm-dicom-docs.on.fleek.co/#/).

jhlegarreta replied:
Done in 894e5e6.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+## Using the Insight Toolkit
+
+### How do I combine ITK and VTK in my application?
```
This is out-of-date and could be removed.

jhlegarreta replied:
Removed in 894e5e6.

thewtex replied:
Still seems to be present in the latest build previews.

jhlegarreta replied:
My bad, Matt. Realized now I did not remove it. Done in 52d18a7 PR #4351.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+## Platform-specific questions
+
+### How do I avoid multiple prompts to reload all projects in Microsoft Visual Studio?
```
out-of-date to be removed

jhlegarreta replied:
Removed in 894e5e6.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+### Is ITK tested?
+
+Please see this page: [ITK/Testing](https://itk.org/Wiki/ITK/Testing).
```
needs refresh

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+Please see the [third party applications](https://itk.org/Wiki/ITK/Third_Party_Applications) page for
+visualization applications that used ITK to perform image
+reading/writing.
```
We should have a third-party applications page.

jhlegarreta replied:
PR #4313.

Missing
https://itk.org/Wiki/ITK/Related_Software
https://itk.org/Wiki/ITK/Developer_Software

from https://itk.org/Wiki/ITK/Other_Software

thewtex reviewed (changes requested):
Great start!

thewtex commented:
> Links to other wiki pages that need to be transitioned:

I added to this issue: https://github.com/InsightSoftwareConsortium/ITK/issues/4063

Those could also be added there so we have one place to keep track of them.

thewtex commented:
>  wiki page deprecations

We could also consider placing a banner on the landing page.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+for which ITK has readers.
+
+A good set of tools for performing such conversions is available in
+ImageMagick, which can be downloaded for free from https://imagemagick.org/index.php.
```
Done in 894e5e6.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+via a factory mechanism based on the file extension
+([`itk::ImageIOFactory`](https://itk.org/Doxygen/html/classitk_1_1ImageIOFactory.html)). Some proprietary files format are only imported.
+More information on [ITK File Formats](https://itk.org/Wiki/ITK_File_Formats) is
+available, including data types supported.
```
Done in 894e5e6.

jhlegarreta replied:
Should eventually add https://wasm.itk.org/en/latest/introduction/file_formats/meshes.html as well?

thewtex replied:
Yes, also transform formats.

jhlegarreta replied:
Meshes/point sets and transforms added in b482b87.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,611 @@
+
+ITK does not provide any visualization functionalities. Applications
+requiring visualization should implement those functionalities from
+internal code or libraries like OpenGL, GLUT or VTK.
```
Removed in 894e5e6.

jhlegarreta commented:
@thewtex Addressed the comments. Others may be willing to chime in to answer the pending questions.

About the links to the Wiki pages that have not been transitioned, I do not think it is feasible to wait until all is transitioned to merge the contributions, as this is a large endeavor.

thewtex commented:
> @thewtex Addressed the comments. Others may be willing to chime in to answer the pending questions.

@jhlegarreta awesome!

> About the links to the Wiki pages that have not been transitioned, I do not think it is feasible to wait until all is transitioned to merge the contributions, as this is a large endeavor.

I agree. If there is outdated content or wiki content, we should omit it for now, though, in my opinion.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+# Frequenty asked questions
```
-> Frequently asked questions (FAQ)

since FAQ is so commonly used?

jhlegarreta replied:
Done in fe3d292.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+([`itk::ImageIOFactory`](https://itk.org/Doxygen/html/classitk_1_1ImageIOFactory.html)). Some proprietary files format are only imported.
+
+More information on supported file formats is available on the
+[wiki ITK File Formats](https://itk.org/Wiki/ITK_File_Formats) and [ITK-wasm Image File Formats](https://wasm.itk.org/en/latest/introduction/file_formats/images.html)
```
I think the wiki link is outdated and should be removed.

jhlegarreta replied:
Removed in fe3d292.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+| Stimulate (spr/sdt) | [`itk::StimulateImageIO`](https://itk.org/Doxygen/html/classitk_1_1TIFFImageIO.html) | |
+| [TIFF](https://en.wikipedia.org/wiki/TIFF) | [`itk::TIFFImageIO`]() | |
+| [Varian FDF](https://github.com/InsightSoftwareConsortium/ITKIOFDF) | | |
+| [VTK Structured Points](https://www.vtk.org/VTK/img/file-formats.pdf) | [`itk::VTKImageIO`](https://itk.org/Doxygen/html/classitk_1_1VTKImageIO.html) | |
```
:sparkler:

jhlegarreta replied:
Resolving.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+[ITK-wasm Mesh and Point Set File Formats](https://wasm.itk.org/en/latest/introduction/file_formats/meshes.html)
+site.
+
+### What trasnform file formats can ITK import and export?
```
trasnform -> transform

jhlegarreta replied:
Fixed in fe3d292.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+   application that allows you to load DICOM series and export them in
+   MetaImage and Analyze format (among others). She graciously has made
+   this application publicly available at:
+   https://darkwing.uoregon.edu/~jolinda/MRIConvert/.
```
Perhaps we should replace with [dcm2niix](https://github.com/rordenlab/dcm2niix)?

jhlegarreta replied:
dcm2niix is great to convert to NIfTI, but not sure it is able to export to MetaImage, and to Analyze. Not about other tools.

thewtex commented on Documentation/docs/index.md:
```diff
@@ -30,6 +30,7 @@ download
 Examples <https://examples.itk.org>
 Tutorial <https://github.com/InsightSoftwareConsortium/ScientificImageAnalysisVisualizationAndArtificialIntelligenceCourse>
 Software Guide <https://itk.org/ItkSoftwareGuide.pdf>
+learn/faq.md
```
We will want to change the `:maxdepth:` above to `1` to address:

![image](https://github.com/InsightSoftwareConsortium/ITK/assets/25432/4c41e4ac-488c-4b00-a10b-3e7b91e3267b)

jhlegarreta replied:
:+1: Done in e808466.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+| File Format | Read/Write | Import |
+| ----------- | ---------- | ------ |
+| [BYU](http://www.eg-models.de/formats/Format_Byu.html) | [
+`itk::BYUMeshIO`]https://itk.org/Doxygen/html/classitk_1_1BYUMeshIO.html | |
```
missing `[`

jhlegarreta replied:
:+1: Added in e808466.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+| InsightLegacy | [`itk::TxtTransformIOTemplate`](https://itk.org/Doxygen/html/classitk_1_1TxtTransformIOTemplate.html) | |
+| ITK HDF5 | [`itk::HDF5TransformIOTemplate`](https://itk.org/Doxygen/html/classitk_1_1HDF5TransformIOTemplate.html) | |
+| MATLAB | [`itk::MatlabTransformIOTemplate`](https://itk.org/Doxygen/html/classitk_1_1MatlabTransformIOTemplate.html) | |
+| MINC (Medical Image NetCDF) | `itk::MINCTransformIOTemplate` | |
```
Missing doxgyen link

jhlegarreta replied:
The module is not build by default, and the Doxygen config does not build all modules, no Doxygen documentation can be linked. Re https://github.com/InsightSoftwareConsortium/ITK/pull/4250#issuecomment-1748754221.

thewtex commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,564 @@
+in case of desperation.
+
+1. Use the [itk-wasm dicom](https://itk-wasm-dicom-docs.on.fleek.co/#/),
+   JavaScript pacakge, which is available from [here](https://itk-wasm-dicom-python-docs.on.fleek.co/).
```
We can remove `, JavaScript packgae,` -- it is a C++ CLI, JavaScript package, and Python package.

jhlegarreta replied:
Removed in e808466.

thewtex reviewed (changes requested):
:rescue_worker_helmet:

jhlegarreta commented:
> Still seems to be present in the latest build previews.

Not sure about the intention. Should that be removed? Maybe reworked to tell people how to build applications with ITK + VTK?

> This paragraph can be removed.

Removed in e808466.

thewtex commented:
>> Still seems to be present in the latest build previews.

> Not sure about the direction of the comment. Should the section be removed? Maybe reworked to tell people how to build applications with ITK + VTK?

Sorry for the lack of clarity. I think I was referencing an additional paragraph that was out-of-date that is gone.

In the future, we could link to the itk-vtk python interface #4320, but there is a lot of great content here. I think we should get it in and start building on it.

thewtex reviewed (approved):
Thank you @jhlegarreta !

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,561 @@
+   [ITKApps/MetaImageImporter](https://github.com/InsightSoftwareConsortium/ITKApps/tree/master/MetaImageImporter).
+   You can also take an existing MetaImage reader and modify it in
+   order to fit your needs. Please find attached one file that you
+   can edit it with your favorite text editor.
```
As per https://github.com/InsightSoftwareConsortium/ITK/pull/4313#discussion_r1390272297, mentions to `ITKApps` should be removed. @thewtex ?

thewtex replied:
Yes, ITKApps is no longer maintained. I do no think we should direct people to it.

jhlegarreta replied:
Done in PR #4364.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,561 @@
+Examples on multimodality registration are available in
+[Examples/RegistrationITKv4](https://github.com/InsightSoftwareConsortium/ITK/tree/master/Examples/RegistrationITKv4). A demo application using mutual
+information is available under
+[ITKApps/LandmarkInitializedMutualInformationRegistration](https://github.com/InsightSoftwareConsortium/ITKApps/tree/master/LandmarkInitializedMutualInformationRegistration).
```
As per https://github.com/InsightSoftwareConsortium/ITK/pull/4313#discussion_r1390272297, mentions to `ITKApps` should be removed. @thewtex ?

thewtex replied:
Yes, ITKApps is no longer maintained. I do no think we should direct people to it.

jhlegarreta replied:
Done in PR #4364.

jhlegarreta commented on Documentation/docs/learn/faq.md:
```diff
@@ -0,0 +1,561 @@
+
+A full tutorial on how to combine ITK and VTK is available at [Getting Started with ITK and VTK](http://www.itk.org/CourseWare/Training/GettingStarted-II.pdf).
+
+You will need the auxiliary classes available at the [ITKApps](https://github.com/InsightSoftwareConsortium/ITKApps)
```
As per https://github.com/InsightSoftwareConsortium/ITK/pull/4313#discussion_r1390272297, mentions to `ITKApps` should be removed. @thewtex ?
